### PR TITLE
Use SledStore to store full timeline

### PR DESCRIPTION
**Pull request by jsparber**
This implements storing and retrieving the full timeline via the `SledStore`.

When a `sync` or `message` response is received the events are written to the store in the same direction (Backward).
The implementation merges consecutive batches as well as overlapping batches into a single batch. For each batch that isn't consecutive the store keeps track of the `prev_batch` and `next_batch` of the batch. If a disconnected batch (a response that isn't possible to be merge to an already known batch) is received a new batch is added to the store and the `prev_batch` and `next_batch` is tracked separately. It may be possible to merge the batches at a later point, but it would require moving the entire batch from one location to an other one inside the store which may be expensive, therefore I would keep them as distinct batches that can grow independently. 

When reading events, the store automatically continues reading events from an other known batch whenever possible. Once the end of the stored timeline is reached, the found events are returned and if not all requested events are found a `token` is added to the response that can be used to retrieve more events from the homeserver.

Please note that this doesn't implement storing the timeline to the `MemoryStore`

Fixes: https://github.com/matrix-org/matrix-rust-sdk/issues/138

### CODE REVIEW

**Comment by jsparber** [comment]
cc: @DevinR528

**Comment by ShadowJonathan** [comment]
While i like the idea, i'd love to know what exactly a good reason would be for the SDK to store the timeline locally.

Technically, the timeline on the server and the timeline "locally" can become out of sync once more DAG events converge and "slot" into historical timelines. When an author expects this to happen, they might want to back-paginate to collect all of these events (somehow) and display them appropriately. Or the author might expect what a server ought to do; display the correct topological ordering of the events on a request.

This is good for caching, but I don't think its good as a Source of Truth, as the historical timeline is subject to change once federation converges, and it's relatively easy for this to happen more than 50 events ago.

Even then, duplicate events are easy to have here as a server might "append" a converged event to a sync, which then gets persisted "late" into the timeline, and when the client might back-paginate to get more events, it might encounter both the persisted "late" variant of the event, and the "correct" variant the server might supply in that moment.

**Comment by jsparber** [comment]
> While i like the idea, i'd love to know what exactly a good reason would be for the SDK to store the timeline locally.
 Requesting events from the server can be slow. Especially for encrypted rooms we need a local stored timeline so that we don't need to decrypt events every time the user views older messages and there is a hard requirement for a local stored timeline for the search in encrypted rooms.
 
 
 Honestly i forgot about about de-duplicating events. I will add it.

**Comment by ShadowJonathan** [comment]
Is there a way to only set the SDK to store some rooms locally? i.e only DMs, like you said.

**Comment by jsparber** [comment]
> Is there a way to only set the SDK to store some rooms locally? i.e only DMs, like you said.

Currently no, it could be added at a later point, but i don't really see a good reason to not store the timeline.

File: matrix_sdk_base/src/store/sled_store/mod.rs [hunk 1]
```diff
@@ -550,6 +648,129 @@ impl SledStore {
...
+                                .map(|e| decode_batch_idx_position(&e).unwrap())
+                            {
+                                let position = position - timeline.events.len();
+                                Some((true, false, batch_idx, position))
+                            } else {
+                                None
+                            }
+                        } else {
+                            None
+                        };
```
Comment (DevinR528) [inline matrix_sdk_base/src/store/sled_store/mod.rs:712]: This works (without checking all four possible combinations) because of storing all events "backwards" right? So you check if the `start` token is in the end `Tree` or if the `end` token is in the start `Tree`?

File: matrix_sdk_base/src/client.rs [hunk 2]
```diff
@@ -949,6 +970,64 @@ impl BaseClient {
...
+        for event in &response.chunk {
+            #[allow(unused_mut)]
+            let mut event: SyncRoomEvent = event.clone().into();
+
+            #[cfg(feature = "encryption")]
+            match hoist_room_event_prev_content(&event.event) {
+                Ok(e) => {
+                    if let AnySyncRoomEvent::Message(AnySyncMessageEvent::RoomEncrypted(
+                        encrypted,
+                    )) = e
```
Comment (DevinR528) [inline matrix_sdk_base/src/client.rs:994]:
```suggestion
                Ok(AnySyncRoomEvent::Message(AnySyncMessageEvent::RoomEncrypted(
                    encrypted,
                )))
```

Maybe a bit easier to read :shrug:

File: matrix_sdk_base/src/store/sled_store/mod.rs [hunk 3]
```diff
@@ -1135,4 +1604,551 @@ mod test {
...
+            .unwrap()
+            .unwrap();
+
+        assert_eq!(backward_events.events.len(), expected_events.len());
+        assert_eq!(backward_events.token, end_token);
+        assert!(!backward_events
+            .events
+            .iter()
+            .zip(expected_events.iter())
+            .any(|(a, b)| a.event_id() != b.event_id()));
```
Comment (DevinR528) [inline matrix_sdk_base/src/store/sled_store/mod.rs:2152]:
```suggestion
        assert!(backward_events
            .events
            .iter()
            .zip(expected_events.iter())
            .all(|(a, b)| a.event_id() == b.event_id()))
```

The double negative was confusing it took me a bit to figure out what was going on haha

Comment (DevinR528) [inline matrix_sdk_base/src/store/sled_store/mod.rs:2152]:
```suggestion
        assert!(backward_events
            .events
            .iter()
            .zip(expected_events.iter())
            .all(|(a, b)| a.event_id() == b.event_id()))
```

The double negative was confusing it took me a bit to figure out what was going on haha. Same for all of the above `! !=`

File: matrix_sdk_base/src/store/sled_store/mod.rs [hunk 4]
```diff
@@ -550,6 +648,129 @@ impl SledStore {
...
+                            .map(|e| decode_batch_idx_position(&e).unwrap())
+                        {
+                            Some((false, true, batch_idx, position))
+                        } else if let Some(end) = &timeline.end {
+                            if let Some((batch_idx, position)) = start_token_to_batch_idx_position
+                                .remove((room.as_str(), end.as_str()).encode())?
+                                .map(|e| decode_batch_idx_position(&e).unwrap())
+                            {
+                                let position = position - timeline.events.len();
+                                Some((true, false, batch_idx, position))
```
Comment (DevinR528) [inline matrix_sdk_base/src/store/sled_store/mod.rs:706]: This works (without checking all four possible combinations) because of storing all events "backwards" right? So you check if the `start` token is in the end `Tree` or if the `end` token is in the start `Tree`?

File: matrix_sdk_base/src/client.rs (hunk 2 above)
Comment (DevinR528) [inline matrix_sdk_base/src/client.rs:994]:
```suggestion
                Ok(AnySyncRoomEvent::Message(AnySyncMessageEvent::RoomEncrypted(
                    encrypted,
                )))
```

Maybe a bit easier to read :shrug:

File: matrix_sdk/src/room/common.rs [hunk 5]
```diff
@@ -105,45 +107,163 @@ impl Common {
...
+    ) -> Result<Option<Vec<SyncRoomEvent>>> {
+        let room_id = self.inner.room_id();
+        let events = if let Some(mut stored) = self
+            .client
+            .store()
+            .get_timeline(room_id, Some(start), end, Some(limit), direction.clone())
+            .await?
+        {
+            // We found a gab or the end of the stored timeline.
+            if let Some(token) = stored.token {
```
Comment (DevinR528) [inline matrix_sdk/src/room/common.rs:166]: Making the request every time is what keeps the cache from ever going stale and it also is what prevents gaps right? It's too bad there isn't a more straightforward way to avoid the request every time.

Could the [/message](https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-client-r0-rooms-roomid-messages) endpoint be used for a more exact finding of the missing events? It would avoid needing to double the limit and could use the given `direction`.

Comment (agraven) [reply matrix_sdk/src/room/common.rs:166]: That comment at 165 should also say "gap" instead of "gab".

File: matrix_sdk_base/src/store/sled_store/mod.rs (hunk 3 above)
Comment (jsparber) [reply matrix_sdk_base/src/store/sled_store/mod.rs:2152]: i didn't see `all` makes much more sense :)

File: matrix_sdk/src/room/common.rs (hunk 5 above)
Comment (jsparber) [reply matrix_sdk/src/room/common.rs:166]: > Making the request every time is what keeps the cache from ever going stale and it also is what prevents gaps right? It's too bad there isn't a more straightforward way to avoid the request every time.

No, the request is made only when not enough events where found in the store.

File: matrix_sdk_base/src/store/sled_store/mod.rs (hunk 4 above)
Comment (jsparber) [reply matrix_sdk_base/src/store/sled_store/mod.rs:706]: exactly, it can only fit before a stored batch or after a stored batch.

File: matrix_sdk/src/room/common.rs (hunk 5 above)
Comment (jsparber) [reply matrix_sdk/src/room/common.rs:166]: > it also is what prevents gaps right?
Not really, but if this hits a gap we fill it with the request. In the specific case that the gap was only less then `limit` we are requesting messages that are actually stored in the store but that's impossible to know before making the request since the size of the gap is only known once it doesn't exist any more, I mean once we filled it.

Comment (jsparber) [reply matrix_sdk/src/room/common.rs:166]: @DevinR528 how would the cache get stale?

Is it possible that the same message requests, with the same two tokens, return different results?

Comment (DevinR528) [reply matrix_sdk/src/room/common.rs:166]: I don't think it is possible for /sync to do that because they send the event in the order they are received but, /messages or /get_context might??? The spec says `in chronological order` but I think that just means small -> big and ordered by topological sort in which case I think a new event could slip in between two old/seen events?

Spec:
> Events are ordered in this API according to the arrival time of the event on the homeserver. This can conflict with other APIs which order events based on their partial ordering in the event graph.

```
a                    b
A ---> B ---> C ---> D
New event comes in from a previously non responsive server
a                            b
A ---> B ---> B' ---> C ---> D
```

**Comment by DevinR528** [comment]
Ugh sorry to clutter everything up with review comments instead of "request changes" reviews that you could actually close. if you what to leave a not on the comment I can delete it when you ready or whatever.

File: matrix_sdk/src/room/common.rs (hunk 5 above)
Comment (jsparber) [reply matrix_sdk/src/room/common.rs:166]: hmm, so the new events (that comes in from a previously non responsive server) are included in a sync response, which will cause a different ordering for the event in the local timeline, but IMO as long as we include all events the specific ordering isn't that important since even different server can have different ordering of the timeline.

**Comment by jsparber** [comment]
> Ugh sorry to clutter everything up with review comments instead of "request changes" reviews that you could actually close. if you what to leave a not on the comment I can delete it when you ready or whatever.

don't really know the difference :)

File: matrix_sdk/src/room/common.rs (hunk 5 above)
Comment (DevinR528) [reply matrix_sdk/src/room/common.rs]: Well the sync response would have a completely set and new token, to continue the example it would be sync response one = `A, B, C, D` as `tkn_1` to `tkn_2` then sync response 2 = `B'` using a new `tkn_3` to `tkn_4` the problem would be if someone used /messages or /get_context, so if we say our `Common::messages` is the same as calling /messages or eventually if we add a `Common::get_context` the cases when the call makes requests and the case where it uses the cache would differ. Having a `start/end` eventId would also be arbitrary (as in they aren't necessarily sequential).

> since even different server can have different ordering of the timeline.

While this may be basically true in practice sort of, the idea is that all servers do agree eventually (eventually consistent).

I do wonder if we just kind of go with what you have (because it is really good IMHO!!) if we could say "our timeline store is ordered the same way sync is ordered for the most part" and as long as we don't have duplicates and keep the chunks together.

**Comment by jsparber** [comment]
i added now deduplication of events. And the removal of redacted events from the store. I think this is good now for a final review.

File: matrix_sdk/src/room/common.rs (hunk 5 above)
Comment (jsparber) [reply matrix_sdk/src/room/common.rs]: Sounds good to me. I also added a method to clear the timeline so that it will be fetched again.

File: matrix_sdk_base/src/store/sled_store/mod.rs [hunk 6]
```diff
@@ -894,15 +1433,50 @@ impl StateStore for SledStore {
...
+    ) -> Result<Option<StoredTimelineSlice>> {
+        self.get_timeline(room_id, start, end, limit, direction).await
+    }
+
+    async fn remove_timeline(&self, room_id: Option<&RoomId>) -> Result<()> {
+        self.remove_timeline(room_id).await
+    }
+}
+
+fn deserialize_event<T: for<'b> Deserialize<'b>>(
```
Comment (poljar) [inline matrix_sdk_base/src/store/sled_store/mod.rs:1453]: Why is this needed? If it is indeed needed does it need to take an owned version of the store key? Doesn't seem so? Can we modify `SledStore::deserialize_event()` so that it uses this method?

**Review by poljar — CHANGES_REQUESTED** [review]
This seems mainly sensible, though a bit of a cleanup of the concepts and splitting up some methods seems like it could make this nicer.

File: matrix_sdk_common/src/deserialized_responses.rs [hunk 7]
```diff
@@ -95,12 +95,26 @@ pub struct SyncRoomEvent {
     pub encryption_info: Option<EncryptionInfo>,
 }
 
+impl SyncRoomEvent {
+    /// Get the event id of this `SyncRoomEvent`
+    pub fn event_id(&self) -> EventId {
+        self.event.get_field::<EventId>("event_id").unwrap().unwrap()
```
[the remark this answers is not in the document]
Comment (jsparber) [reply matrix_sdk_common/src/deserialized_responses.rs:101]: what do you mean? why does it not work?

[the remark this answers is not in the document]
Comment (poljar) [reply matrix_sdk_common/src/deserialized_responses.rs:101]: The unwraps need to be handled.

[the remark this answers is not in the document]
Comment (jsparber) [reply matrix_sdk_common/src/deserialized_responses.rs:101]: aren't we sure at this point that the event is valid? And if it's a valid event we can be sure that it includes the `event_id`

[the remark this answers is not in the document]
Comment (poljar) [reply matrix_sdk_common/src/deserialized_responses.rs:101]: We aren't, we convert the event over here: https://github.com/matrix-org/matrix-rust-sdk/blob/master/matrix_sdk_base/src/client.rs#L446 and then we unconditionally push it into the custom timeline struct over here: https://github.com/matrix-org/matrix-rust-sdk/blob/master/matrix_sdk_base/src/client.rs#L548.

File: matrix_sdk/src/room/common.rs [hunk 8]
```diff
@@ -105,45 +108,177 @@ impl Common {
...
     /// ```
     pub async fn messages(
         &self,
-        request: impl Into<get_message_events::Request<'_>>,
-    ) -> Result<get_message_events::Response> {
-        let request = request.into();
-        self.client.send(request, None).await
+        start: Option<&EventId>,
+        end: Option<&EventId>,
+        limit: usize,
```
[the remark this answers is not in the document]
Comment (jsparber) [reply matrix_sdk/src/room/common.rs:205]: i changed it to `u32`, but I kept using usize in the store.

File: matrix_sdk_common/src/deserialized_responses.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jsparber) [reply matrix_sdk_common/src/deserialized_responses.rs:101]: Can I just drop events that don't have a valid `event_id` and not store them in the store?

[the remark this answers is not in the document]
Comment (poljar) [reply matrix_sdk_common/src/deserialized_responses.rs:101]: Even if you drop them from the store, this method can be used for events that aren't part of the store and thus can't panic if the event id is missing.

Dropping them from the store sounds sensible, considering that we'll likely using the event id as part of the key to store the value.

File: matrix_sdk_base/src/store/sled_store/mod.rs [hunk 9]
```diff
@@ -550,6 +649,182 @@ impl SledStore {
...
+                                    .map(|t| decode_batch_idx_position(&t).unwrap())
+                                {
+                                    if let Some(full_event) = timeline_events
+                                        .get((room.as_str(), batch_idx, position).encode())?
+                                        .map(|t| {
+                                            self.deserialize_event::<SyncRoomEvent>(&t).unwrap()
+                                        })
+                                        .and_then(|e| e.event.deserialize().ok())
+                                    {
+                                        if let Ok(Some(room_version)) = self.get_room_version(room)
```
[the remark this answers is not in the document]
Comment (jsparber) [reply matrix_sdk_base/src/store/sled_store/mod.rs:781]: Not sure but redaction changes based on the room version and different values are stripped. From reading the specs it looks like only `m.room.aliases` has changed in version 6. Before version 6 it allowed to keep `aliases` after redaction. I think the room version doesn't have any implication for the store, but we may not strip the correct data or to much data.

File: matrix_sdk_base/src/store/sled_store/mod.rs (hunk 6 above)
Comment (jsparber) [reply matrix_sdk_base/src/store/sled_store/mod.rs:1473]: hmm, rustc didn't like to move self into a closure. Don't remember what the issue was. I will remove the additional function.

**Comment by ftilde** [comment]
First of all: Thank you for working on this feature. For implementing a chat client this is really useful, since otherwise this functionality has to be implemented in the client itself. For this reason I've actually been using this PR for a couple of days. I've stumbled upon a bug, though:

### Bug Description

The timeline-handling part of `save_changes` in SledStore assumes that the new timeline slices intersects with at most one slice/batch already in the store. This is (in the current implementation) not the case. Assume A and B are batches that (for example) have been added to the store via a sync. Since then, another sync has not happened, but not added new event batches to the store. Hence, the sync token has changed. When we now fetch a large batch of messages using the latest sync token, this batch may overlap multiple batches already in the store:

```
|   A   |----|   B   |----
|         C       |
```

The current code in `save_changes` merges A and C, i.e., it adds events that are in C, but not in A to A and sets the new end token to that of C.
Crucially, all events that are in C, but also in B are _not_ added to A, but the end token _still_ points halfway into B:

```
Events:
|   A        |
             |   B   |

start/end-tokens:
|   A             |
             |   B   |
```

As a result, when we later get messages from the store, (1) the end token of A is not linked to B in any way, which results in another fetch from the end of A and (2) the combined result of A and the fetched batch will NOT contain the events that are in B, but not in A!

To make things worse, this situation is really common:
All events that arrive in the store from syncs do so in batches that are not linked (at least most of the time).
For example, here are the start/end tokens of two messages that arrived in short succession via syncs:

```
message 1
start:  s3794_396052_11800_6878_862_3_1972_9819_1
end:    s3793_396052_11800_6878_862_3_1972_9819_1

message 2
start:  s3795_396059_11802_6878_862_3_1972_9819_1
end:    s3794_396059_11802_6878_862_3_1972_9819_1
```

### Possible Fixes

I have two suggestions for fixing this, but I'm not completely sure which is to prefer:

1. Avoid explicitly fetching event batches that intersect with more than one batch already in the store. When fetching the newest messages that could be done by including the end token of the newest batch in the store as the end parameter in the request.
Since the newest batch in the store may possibly contain less than `limit` messages, this would also need a second call to `get_timeline` after the fetch has succeeded: When we limit the requested batch like this, it may be shorter than `limit` messages.

2. Remove the limitation of `save_changes` that only two batches can be merged and instead merge all batches that intersect with the new timeline slice. I don't think this would be much more expensive than the current approach, since we already check for all events if they are part of some stored batch already. I'm not sure if implementing the first suggestion would fix all cases where a situation like this can occur, I would personally vote for this option. It's just more robust, also considering future changes.

File: matrix_sdk_base/src/store/sled_store/mod.rs [hunk 10]
```diff
@@ -550,6 +649,182 @@ impl SledStore {
...
+                            // Handle overlapping slices.
+                            let first_event = timeline.events.first().unwrap().event_id();
+                            let last_event = timeline.events.last().unwrap().event_id();
+
+                            let found_first = if let Some((batch_idx, current_position)) =
+                                event_id_to_position
+                                    .get((room.as_str(), first_event.as_str()).encode())?
+                                    .map(|e| decode_batch_idx_position(&e).unwrap())
+                            {
+                                // Check if next event in store fits next event
```
[the remark this answers is not in the document]
Comment (jsparber) [reply matrix_sdk_base/src/store/sled_store/mod.rs]: i removed it.

**Comment by jsparber** [comment]
@poljar i fixed all requested changes. Can you have a second look?

**Comment by jsparber** [comment]
@ftilde sounds like a reasonable issue. I think your second option sounds like a better solution. I will need to do some thinking how to implement it though.

**Comment by poljar** [comment]
> @poljar i fixed all requested changes. Can you have a second look?

Sure, will do.

**Comment by jsparber** [comment]
@ftilde can you report if https://github.com/matrix-org/matrix-rust-sdk/pull/288/commits/5425f2f29ad50c37c8f15edb1594e9309c0804e5 fixes the issue. Now instead of ignoring new events we know already the previously stored events are removed. The only downside to this is that we know can have "invalid" events in the store and we may have empty batches.

**Comment by jsparber** [comment]
@poljar the ci fails again for issues outside my code changes.

**Comment by poljar** [comment]
> @poljar the ci fails again for issues outside my code changes.

That looks like something went wrong with the fresh Ruma releases? @jplatte any idea if that's something actually wrong with Ruma?

**Comment by jplatte** [comment]
Yeah I did make a stupid mistake, but immediately yanked the affected versions. Let me check what's going on.

**Comment by jplatte** [comment]
Oh, the ruma-api release broke some of the ruma-*-api crates. I guess I need to yank yet again.

**Comment by poljar** [comment]
> Done. I guess I just need to go for another breaking-change release soon.

Thanks. CI has been restarted.

**Comment by jsparber** [comment]
@poljar @jplatte thanks :)

**Comment by ftilde** [comment]
@jsparber Hm, so the patch actually seemingly fixes the incorrect behavior that I observed and that led me to investigate. However, I think that it does not generally fix the problem and in particular is still broken in the example from above.

In this situation...
```
|   A   |----|   B   |----
|         C       |
```
... now all events from C get absorbed into A and removed from B. But the start token of B is unchanged, isn't it?

```
Events:
|   A             |
                  |B |

start/end-tokens:
|   A             |
             |   B   |
```
If we now fetch starting from an event in B and go forward, we will skip all the events that previously were in B, but were now added to A.
Specifically, in `sled_store/mod.rs:1242` and following lines we would take the first if-let branch (with the incorrect start token of B), but ignore the second (since there is no matching linked batched) and thus break and fetch from the server using the (incorrect) start token.

(I may be wrong as I've only mentally stepped through the code with this example in mind. Maybe we can write a test for this? Not sure how easy that is in this particular case.)

If we don't want to proactively merge A and B in the example above, we will have to "link" them properly with the start/end tokens, I think. This will reduce superfluous server requests as well. I might be able to help and think about how exactly to do this in a couple of days, if you like.

**Comment by jsparber** [comment]
@ftilde i think your  observation is correct. I think it would be enough to update the start/end token of the batch we removed events from. Would be nice to have a test case for this specific issue.

**Comment by jsparber** [comment]
actually we can't just update the start/end token, since we can't be completely sure whether events are duplicated or an overlap is happening. Not sure how to solve this issue.

**Comment by ShadowJonathan** [comment]
@jsparber this is exactly what i mentioned in my [first comment](https://github.com/matrix-org/matrix-rust-sdk/pull/288#issuecomment-869894572);

> Even then, duplicate events are easy to have here as a server might "append" a converged event to a sync, which then gets persisted "late" into the timeline, and when the client might back-paginate to get more events, it might encounter both the persisted "late" variant of the event, and the "correct" variant the server might supply in that moment.

And i still stand by that, I think timeline caching is a bad idea.

**Comment by ftilde** [comment]
@jsparber Just an idea which I haven't thought through completely: It seems the more complicated of those two problems seems to be the duplication of events (and in general: the arrival-time-based ordering of events from syncs). How about only caching messages/events from explicit calls to this API? The [spec for room/messages](https://matrix.org/docs/spec/client_server/latest#get-matrix-client-r0-rooms-roomid-messages] does not mention inconsistent/changed ordering of events. This would have the added benefit of not increasing the store size for clients (e.g., bots) which do not use this API.

**Comment by jsparber** [comment]
> The spec for room/messages does not mention inconsistent/changed ordering or duplication of events. This would have the added benefit of not increasing the store size for clients (e.g., bots) which do not use this API.

This is also the case. I don't think that the event sequence  changes but new events could be added, since the arrived at a later point at the homeserver.

File: matrix_sdk/src/room/common.rs [hunk 11]
```diff
@@ -137,17 +144,217 @@ impl Common {
...
+
+    async fn request_missing_messages(
+        &self,
+        mut stored: StoredTimelineSlice,
+        end: Option<&EventId>,
+        limit: u32,
+        direction: &Direction,
+    ) -> Result<Vec<SyncRoomEvent>> {
+        if let Some(token) = stored.token {
+            trace!("A gap in the stored timeline was found. Request messages from token {}", token);
```
Comment (poljar) [inline matrix_sdk/src/room/common.rs:249]:
```suggestion
            trace!(token=token, "A gap in the stored timeline was found. Requesting messages from the server.");
```

File: matrix_sdk/src/room/common.rs [hunk 12]
```diff
@@ -137,17 +144,217 @@ impl Common {
...
+        direction: &Direction,
+    ) -> Result<Option<Vec<SyncRoomEvent>>> {
+        let start = if let Some(start) = start {
+            start
+        } else {
+            // If no start event was given it's not possible to find the context of it.
+            return Ok(None);
+        };
+
+        trace!("The start event with id {} wasn't found in the stored timeline. Fallback to the context API", start);
```
Comment (poljar) [inline matrix_sdk/src/room/common.rs:307]: Same as above, pass the start as an argument to `trace!()`.

File: matrix_sdk_base/src/client.rs [hunk 13]
```diff
@@ -959,6 +980,59 @@ impl BaseClient {
...
+                    warn!("Error deserializing event {:?}", error);
+                }
+            }
+
+            events.push(event);
+        }
+
+        let (chunk, start, end) = match direction {
+            Direction::Backward => {
+                (events.clone(), response.start.clone().unwrap(), response.end.clone())
```
Comment (poljar) [inline matrix_sdk_base/src/client.rs:1019]: The unwrap either need to be gone or, if we're sure that it's ok, turned into an except.

File: matrix_sdk_base/src/client.rs [hunk 14]
```diff
@@ -959,6 +980,59 @@ impl BaseClient {
...
+            }
+            Direction::Forward => (
+                events.iter().rev().cloned().collect(),
+                response.end.clone().unwrap(),
```
Comment (poljar) [inline matrix_sdk_base/src/client.rs:1023]: Same as above, no unwraps please.

**Review by poljar — CHANGES_REQUESTED** [review]
Looking better though there are still a bunch of unwraps that make me uneasy and more importantly fetching events one by one doesn't work since you always get the same event that you use as `start`.

**Comment by agraven** [comment]
Just in case you haven't noticed, these are the typos the CI is complaining about
```
$ typos .
error: `enought` should be `enough`
  --> ./crates/matrix-sdk-base/src/store/sled_store/mod.rs:1307:23
     |
1307 |                 // Not enought events where found in this batch, therefore, go to the previous
     |                        ^^^^^^^
     |
error: `enought` should be `enough`
  --> ./crates/matrix-sdk-base/src/store/sled_store/mod.rs:1388:23
     |
1388 |                 // Not enought events where found in this batch, therefore, go to the previous
     |                        ^^^^^^^
     |
error: `begining` should be `beginning`
  --> ./crates/matrix-sdk-common/src/deserialized_responses.rs:269:46
    |
269 |     /// If `None` this `TimelineSlice` is the begining of the room
    |                                               ^^^^^^^^
    |
error: `recevied` should be `received`
  --> ./crates/matrix-sdk/src/room/common.rs:259:29
    |
259 |             // FIXME: we may recevied an invalied server response that ruma considers valid
    |                              ^^^^^^^^
    |
```

File: crates/matrix-sdk-base/src/store/sled_store/mod.rs [hunk 15]
```diff
@@ -168,6 +196,67 @@ pub fn decode_key_value(key: &[u8], position: usize) -> Option<String> {
...
+    }
+}
+
+impl BatchIdx {
+    fn next(self) -> Self {
+        Self(self.0 + 1)
+    }
+
+    fn to_be_bytes(self) -> [u8; 8] {
+        self.0.to_be_bytes()
```
Comment (Newbytee) [inline crates/matrix-sdk-base/src/store/sled_store/mod.rs:216]: This only works on 64-bit architectures. Since `self.0` here is a `usize`, the `[u8; 8]` return type will be invalid as `self.0.to_be_bytes()` returns `[u8; 4]` on 32-bit architectures.

**Comment by ftilde** [comment]
After the previous discussion, I fear that implementing a local message store is (at least almost) useless.
I've tried to summarize my thoughts below (Sorry for doing that only now. I've been quite busy in the last weeks.)
Of course, feel free to disagree! This is based on my current (possibly flawed) understanding of the ordering of messages returned by syncs and `messages`: As far as I understand there are no guarantees whatsoever about the ordering of one vs. the other.

Let the following be the "true" timeline, i.e. a sequence of events with the ordering decided by the server (after reordering, what would returned by `messages` at a time infinitely far in the future.)

```
most recent <-> older
A001..A500 B001..B500 C D001..D500
```

Then, let this be the ordering in which events *arrive* at (your) homeserver.
Notice that event C is delayed significantly for some reason (temporary disconnection of the homeserver of the user who sent `C` or something like that).

```
most recent <-> older
A001..A500 C B001..B500 D001..D500
```

Now assume that your client is running when D500 arrives, but closed after receiving B001 via sync.
`B001-D500` are now in the local store.
In particular, it did *not* receive `C` and is thus not in the store.

After `A001..A500 C` have arrived at the (your) homeserver, the client is restarted.
Some events `A001..A???` are recevied at the first sync.
Events `A???(+1)..A500` can be received via the sync token and calls to `messages`.
After `A500`, `B001..BXYX` are returned via `messages`, but (due to the large number of messages between `B001` and `C`) `C` is very likely not in the batch.
What should the client do now? Since B001 is already in the store it probably does not make sense to query further messages.
But then it never displays C! The only way to make sure to receive _all_ messages is to re-fetch all messages until the beginning of the timeline.
This behavior, however, would make storing the messages useless. (Or at least almost useless: The single benefit I can think of would be to have an offline copy of a once accurate timeline.)

@ShadowJonathan does this describe your concerns expressed above as well?

**Comment by ShadowJonathan** [comment]
Yes, as currently most of this behaviour is undefined in the spec, all homeservers take different approaches with the sync, and you cannot have a guaranteed ordering story like that. I saw that coming when I first saw this PR, so I still really recommend against merging it.

At most, an event-id -> content cache would work, not a timeline store like this.

**Comment by agraven** [comment]
While do agree that the raised issue is a concern to be considered, my main concern is that not merging this is going to lead to library consumers implementing their own implementation of this feature, which in most cases won't be put through as much scrutiny as an implementation in the matrix-sdk itself, and additionally won't be able to draw on any of the benefits of access to internals that come from being part of the matrix-sdk crate itself.

**Comment by ftilde** [comment]
I think what could be implemented with the current state of spec and 
implementations would be an in memory cache of messages received via 
sync and `messages`. It would have to be invalidated/thrown away 
whenever a message in a sync is missed. That would happen due to a large 
number of messages arriving in a short period, which would be very 
unlikely, however, I think. Such a gap would be detected by comparing 
the start/end tokens of two successive syncs, which should be the same 
if no messages were missed.

**Comment by ShadowJonathan** [comment]
Keep in mind that messages can also be missed if the "limited"/"partial" flag was set (can't remember the exact name), which highlights a partial sync (for a particular room's timeline)

**Comment by poljar** [comment]
> Such a gap would be detected by comparing the start/end tokens of two successive syncs, which should be the same if no messages were missed.

> Keep in mind that messages can also be missed if the "limited"/"partial" flag was set (can't remember the exact name), which highlights a partial sync (for a particular room's timeline)

Those two are the same case, no? Either there is a gap and the limited flag is set *and* the tokens differ, or there isn't a gap in which case the limited flag is not set *and* the tokens match.

**Comment by ShadowJonathan** [comment]
I believe https://github.com/matrix-org/matrix-doc/issues/3263 hasn't been mentioned yet, but it is relevant to this PR

**Comment by jsparber** [comment]
With the current [limits](https://github.com/matrix-org/matrix-doc/issues/3280) of the sync api, I think we can only implement a cache for a continues stream of events of the timeline. This will make the cache a little less efficient, but unfortunately I don't see any way around this. On duplicated events the local cache is dropped, because it's the indication that the order of the timeline on the server has changed. Till the developer requests a new stream we will keep the local order of events, and thus, we can just drop the second occurrence of the event. 

For now I'm not considering implementing any cache for requests directly to [get_message_events](https://docs.ruma.io/ruma/api/client/r0/message/get_message_events/index.html#),  [get_context](https://docs.ruma.io/ruma/api/client/r0/context/get_context/index.html#) and [get_room_event](https://docs.ruma.io/ruma/api/client/r0/room/get_room_event/index.html#). Even though i think it could be possible to serve some requests locally quite easily.

In specific this will solve the following issues we faced so far in this MR:
- No gaps in the returned and stored timeline
- No issue with duplicated events (since the second occurrence of the events can be dropped from the stream)
- No overlapping batches, because we always use the proper token from the previous request

The local cache is invalidated on:
- A sync response contains a limited timeline [1]
- We found duplicated events in the previously started stream. This ensures that we get the new and proper order of events next time the stream is started (the previously started stream won't be affected)

This will still give us a local cache of the timeline for:
- Timeline before receiving a sync response (after the initial sync)
- Offline use
- Quicker loading time on cache hit.
- Developer doesn't need to mess around with batch token

For API, i think we could use [DoubleEndedStream](https://docs.rs/async-std/latest/async_std/stream/trait.DoubleEndedStream.html#). It's still an unstable feature, therefore, I would implement it by using two [Stream](https://docs.rs/futures/latest/futures/stream/trait.Stream.html#), one forward stream and one backward stream starting from the current point in the timeline.

[1] Considerations about filling the gap: 
In future we could try to request events to fill the gap but we still need to invalidate the cache on duplicated events. At least, we wouldn't have any issue with overlapping batches since [get_message_events](https://docs.ruma.io/ruma/api/client/r0/message/get_message_events/index.html#) allows us to set the end token.

**Comment by jsparber** [comment]
closing in favor of #486

**Comment by dkasak** [comment]
@jsparber, you seem to have linked to the wrong thing here. Could you fix the link?
